File: unity/3DBuilderRuntime/Assets/Scripts/Protocol/ProtocolConstants.cs
using System;

namespace ThreeDBuilder.Protocol
{
    /// <summary>
    /// Supported protocol versions for Flutter ↔ Unity communication.
    /// </summary>
    public static class ProtocolVersion
    {
        public const string V1 = "1.0";

        /// <summary>
        /// Returns true if the given version string is a supported protocol version.
        /// </summary>
        public static bool IsSupported(string version)
        {
            return version == V1;
        }
    }

    /// <summary>
    /// Commands sent from Flutter → Unity.
    /// Wire values use snake_case to match the Dart-side protocol constants.
    /// </summary>
    public enum EngineCommand
    {
        Initialize,
        LoadScene,
        ClearScene,
        Dispose
    }

    /// <summary>
    /// Events emitted from Unity → Flutter.
    /// Wire values use snake_case to match the Dart-side protocol constants.
    /// </summary>
    public enum EngineEventType
    {
        Initialized,
        SceneLoading,
        SceneReady,
        Error,
        PerformanceStats
    }

    /// <summary>
    /// Conversion helpers between enum values and their snake_case wire strings.
    /// </summary>
    public static class ProtocolEnumExtensions
    {
        // ── EngineCommand ────────────────────────────────────────────────

        public static string ToWireValue(this EngineCommand command)
        {
            switch (command)
            {
                case EngineCommand.Initialize:  return "initialize";
                case EngineCommand.LoadScene:    return "load_scene";
                case EngineCommand.ClearScene:   return "clear_scene";
                case EngineCommand.Dispose:      return "dispose";
                default:
                    throw new ArgumentOutOfRangeException(nameof(command), command, "Unknown command");
            }
        }

        /// <summary>
        /// Parses a wire-format string to an EngineCommand.
        /// Returns false if the value is unknown.
        /// </summary>
        public static bool TryParseCommand(string wireValue, out EngineCommand command)
        {
            switch (wireValue)
            {
                case "initialize":  command = EngineCommand.Initialize;  return true;
                case "load_scene":  command = EngineCommand.LoadScene;   return true;
                case "clear_scene": command = EngineCommand.ClearScene;  return true;
                case "dispose":     command = EngineCommand.Dispose;     return true;
                default:            command = default;                    return false;
            }
        }

        // ── EngineEventType ──────────────────────────────────────────────

        public static string ToWireValue(this EngineEventType eventType)
        {
            switch (eventType)
            {
                case EngineEventType.Initialized:      return "initialized";
                case EngineEventType.SceneLoading:     return "scene_loading";
                case EngineEventType.SceneReady:       return "scene_ready";
                case EngineEventType.Error:            return "error";
                case EngineEventType.PerformanceStats: return "performance_stats";
                default:
                    throw new ArgumentOutOfRangeException(nameof(eventType), eventType, "Unknown event type");
            }
        }
    }
}

File: lib/engine/protocol/protocol_constants.dart
/// Protocol version and type-safe enums for the Flutter ↔ Unity wire format.
///
/// All commands sent to Unity and events received from Unity must use
/// these constant values. Unknown commands or events must be rejected.
class ProtocolVersion {
  ProtocolVersion._();

  /// The current protocol version used for all envelopes.
  static const String v1 = '1.0';

  /// The set of all valid protocol versions.
  static const Set<String> supported = {v1};
}

/// Commands sent from Flutter → Unity.
enum EngineCommand {
  initialize('initialize'),
  loadScene('load_scene'),
  clearScene('clear_scene'),
  dispose('dispose');

  const EngineCommand(this.value);

  /// The wire-format string sent over the platform channel.
  final String value;

  /// Looks up a command by its wire value. Returns null if unknown.
  static EngineCommand? fromValue(String value) {
    for (final cmd in values) {
      if (cmd.value == value) return cmd;
    }
    return null;
  }
}

/// Events received from Unity → Flutter.
enum EngineEventType {
  initialized('initialized'),
  sceneLoading('scene_loading'),
  sceneReady('scene_ready'),
  error('error'),
  performanceStats('performance_stats');

  const EngineEventType(this.value);

  /// The wire-format string received from the platform channel.
  final String value;

  /// Looks up an event type by its wire value. Returns null if unknown.
  static EngineEventType? fromValue(String value) {
    for (final evt in values) {
      if (evt.value == value) return evt;
    }
    return null;
  }
}
